### PR TITLE
[#1710] Don't touch the project again if it was already built

### PR DESCRIPTION
**Pull request by szarnekow**
The window of opportunity for the autobuild is the time between
forgetting the last built state and the touch of the project in its own
job. Check if there is a known built state before touching the project
again.

Signed-off-by: Sebastian Zarnekow <sebastian.zarnekow@gmail.com>

### CODE REVIEW

**Comment by szarnekow** [comment]
cc @Bananeweizen Can you please give this a try?

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuilderStateDiscarder.java [hunk 1]
```diff
@@ -34,12 +37,34 @@
 public class BuilderStateDiscarder {
 
 	private static final Logger logger = Logger.getLogger(BuilderStateDiscarder.class);
+	private static final MethodHandle InternalBuilder_getLastBuiltTree;
 
+	static {
+		try {
+			@SuppressWarnings("restriction")
+			MethodHandles.Lookup lookup = MethodHandles.privateLookupIn(
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuilderStateDiscarder.java:45]: i assume if we keep java 8 we need to do similar old school reflection as in JavaBuilderState / BuildManagerAccess

**Comment by cdietrich** [comment]
mmm on the first try it does not work

**Comment by szarnekow** [comment]
Testing this is indeed challenging. I’ll revisit the fix.

**Comment by rytina** [comment]
>do you have normal project or java projects?
@cdietrich @szarnekow we have normal projects. Our Xtext grammar does not use Xbase.

**Comment by szarnekow** [comment]
> > do you have normal project or java projects?
> 
> @cdietrich @szarnekow we have normal projects. Our Xtext grammar does not use Xbase.

Thank you for the quick response. The approach taken in the current state of the merge request is not sufficient yet for either type of projects. There are more moving parts involved and an unpleasant mix of different locks at play that need to be straighten. I’ll keep you posted.

**Comment by cdietrich** [comment]
Bug486584Test fails even if I replace the jar with a jar containing a buildertestlanguage file `namespace viech { object Viech }`
and the test file to `namespace bar { object B references viech.Viech }`

**Comment by szarnekow** [comment]
Let's see what Jenkins thinks about this approach.

**Comment by szarnekow** [comment]
@Bananeweizen @rytina  @cdietrich I think this is good for a review and some battle testing.

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuilderStateDiscarder.java [hunk 2]
```diff
@@ -33,50 +29,42 @@
...
-
 	/**
 	 * Returns true, if the given builderArguments indicate that we should discard the built state
 	 * for the given projects.
 	 */
 	public boolean forgetLastBuildState(Iterable<IProject> toUpdate, Map<String, String> builderArguments) {
+		/*
+		 * Implementation note:
+		 * This is an unsafe operation. There no guarantee that any lock is being acquired by the current thread
+		 * while performing this operation. It is  very well possible that this is running concurrently with a build.
```
Comment (tivervac) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuilderStateDiscarder.java:40]: There's a double space after `It is`

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuilderStateDiscarder.java [hunk 3]
```diff
@@ -33,50 +29,42 @@
...
+		 * 
+		 * When we come here, we might, or might not
+		 * - hold the AbstractBuilderState.loadLock
+		 * - hold the org.eclipse.core.internal.resources.WorkManager.lock
+		 * - come from a build and see a situation with org.eclipse.core.internal.resources.Workspace.treeLocked being the case
```
Comment (tivervac) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/BuilderStateDiscarder.java:45]: There's two spaces at the end here

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 4]
```diff
@@ -158,20 +159,141 @@ public BuilderPreferences () {
...
+		}
+	}
+	
+	/**
+	 * True if the next regular build (not a clean build) that will be executed, will be treated as a full build request.
+	 * 
+	 * @since 2.26
+	 */
+	private boolean nextBuildIsFullBuild = false;
+	/**
```
Comment (tivervac) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java:196]: Should be a empty newline before this line

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 5]
```diff
@@ -198,16 +320,18 @@ public boolean isCanceled() {
...
 			} else {
 				IResourceDelta delta = getDelta(getProject());
 				if (delta == null || isOpened(delta)) {
-					fullBuild(progress.split(1), IBuildFlag.RECOVERY_BUILD.isSet(args));
+					fullBuild(progress.split(1), !wasFullBuildRequested && IBuildFlag.RECOVERY_BUILD.isSet(args));
 				} else {
 					incrementalBuild(delta, progress.split(1));
 				}
 			}
+			// If no exceptions have been thrown, fix the information about the requested full builds
```
Comment (tivervac) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java:333]: Just trying to clarify the scenario.

IIUC, we would call `requestFullBuild`. This will trigger an autobuild. The autobuild will then somewhere along the line call this method.

In this case we have `nextBuildIsFullBuild` be true here and `allowRequestFullBuild` false. If an exception did now occur, we couldn't request a new full build, the next build will be a full build, but there currently (possibly) is no more build.

In short, if an exception occurred, it's possible that we have an unbuilt project that we do not retry to build.

**Comment by cdietrich** [comment]
does not work for mydsl general projects for me

```
[Worker-0: Building] Building demo2
[Worker-0: Building] Built demo2 in 10 ms
[Worker-7: Building] interrupted
[Worker-7: Building] Build interrupted.
[Worker-7: Building] Built demo2 in 1045 ms

```

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java (hunk 5 above)
Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java:333]: The AutoBuildJob will be rescheduled in case of cancellation.

**Comment by cdietrich** [comment]
same on open a closed mydsl general project after restart
see a interruped build only and not a follow up build

**Comment by cdietrich** [comment]
looks like the problem i saw happens on all opens.
TODO: check if works in real project
TODO: check if old behaviour can easily be restored by projects when needed (subclassing)

**Comment by rytina** [comment]
@szarnekow I copied the changed files over to our target to test the bug locally and I can confirm it's fixed. The builds only happens once and the recovery build is also working perfectly. Thank you! :-)

**Review by rytina — APPROVED** [review]

**Comment by rytina** [comment]
@szarnekow could you please merge your PR and provide us an update site, so that I can test your change with all other changes that will come with Xtext 2.26?
